### PR TITLE
Add regrets reporter summary table (fixes #2572)

### LINKED ISSUE #2572: Create derived dataset for youtube regrets reporter
As discussed on slack, we want a dashboard to provide summary metrics for the RegretsReporter data.  The raw data is restricted access, so we want to run an ETL to create a summary table accessible to anyone in Mozilla.  The query should be:

```sql
WITH base_t AS (
    SELECT
        metrics.string.metadata_installation_id AS installation_id,
        DATE(submission_timestamp) as date,
        ANY_VALUE(metrics.string.metadata_experiment_arm) AS experiment_arm,
        ANY_VALUE(metrics.string.metadata_feedback_ui_variant) AS ui_arm,
        ANY_VALUE(metadata.geo.country) AS country,
        ANY_VALUE(metadata.user_agent.browser) AS browser,
        IF(LOGICAL_OR(e.name = "video_recommended"), 1, 0) + (IF(LOGICAL_OR(e.name = "regret_action"), 1, 0) * 2) + (IF(LOGICAL_OR(e.name = "video_played"), 1, 0) * 4) AS activities,
    FROM
        `moz-fx-data-shared-prod.regrets_reporter_ucs_live.main_events_v1`,
        UNNEST(events) e
    WHERE
        submission_timestamp >= "2021-12-2"
    GROUP BY
        metrics.string.metadata_installation_id,
        DATE(submission_timestamp)
),
dau_t AS (
SELECT
    COUNT(DISTINCT installation_id) AS dau,
    date,
    experiment_arm,
    ui_arm,
    country,
    browser,
    activities
FROM
    base_t
GROUP BY
    date,
    experiment_arm,
    ui_arm,
    country,
    browser,
    activities
ORDER BY
    date
),
dates_t AS (
    SELECT
        date AS d
    FROM
        dau_t
    GROUP BY
        date
),
wau_t AS (
    SELECT
        COUNT(DISTINCT installation_id) AS wau,
        d AS date,
        experiment_arm,
        ui_arm,
        country,
        browser,
        activities
    FROM
        dates_t 
    INNER JOIN
        base_t
    ON date BETWEEN DATE_SUB(dates_t.d, INTERVAL 6 DAY) AND dates_t.d
    GROUP BY
        d,
        experiment_arm,
        ui_arm,
        country,
        browser,
        activities
    ORDER BY
        d
),
new_user_base_t AS (
  SELECT
    metrics.string.metadata_installation_id AS installation_id,
    MIN(DATE(submission_timestamp)) as date,
    ANY_VALUE(metrics.string.metadata_experiment_arm) AS experiment_arm,
        ANY_VALUE(metrics.string.metadata_feedback_ui_variant) AS ui_arm,
        ANY_VALUE(metadata.geo.country) AS country,
        ANY_VALUE(metadata.user_agent.browser) AS browser,
        IF(LOGICAL_OR(e.name = "video_recommended"), 1, 0) + (IF(LOGICAL_OR(e.name = "regret_action"), 1, 0) * 2) + (IF(LOGICAL_OR(e.name = "video_played"), 1, 0) * 4) AS activities,
  FROM
    `moz-fx-data-shared-prod.regrets_reporter_ucs_live.main_events_v1`,
    UNNEST(events) e
  WHERE
    submission_timestamp >= "2021-12-2"
  GROUP BY
    installation_id
),
new_user_t AS (
    SELECT
        COUNT(DISTINCT installation_id) AS new_users,
        date,
        experiment_arm,
        ui_arm,
        country,
        browser,
        activities
    FROM
        new_user_base_t
    GROUP BY
        date,
        experiment_arm,
        ui_arm,
        country,
        browser,
        activities
    ORDER BY
        date
)
    SELECT 
        dau,
        wau,
        new_users,
        date,
        experiment_arm,
        ui_arm,
        country,
        browser,
        activities
    FROM
        dau_t
    FULL JOIN
        wau_t
    USING(date, experiment_arm, ui_arm, country, browser, activities)
    FULL JOIN
        new_user_t 
    USING(date, experiment_arm, ui_arm, country, browser, activities)
```

### PR DESCRIPTION
**Pull request by kik-kik**
# Add regrets reporter summary table (fixes #2572)

Fixes #2572

Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title)
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated

### CODE REVIEW

**Comment by kik-kik** [comment]
Ran the following commands:
```bash
./bqetl dag create bqetl_regrets_reporter_summary --schedule-interval "0 4 * * *" --owner kignasiak@mozilla.com --description "Measure usage of the regrets reporter addon" --start-date 2021-12-12
./bqetl query schedule regrets_reporter_derived.regrets_reporter_summary_v1 --dag bqetl_regrets_reporter_summary --task-name regrets_reporter_summary__v1
./bqetl dag generate bqetl_regrets_reporter_summary
```

**Comment by kik-kik** [comment]
Both `ci/circleci: dry-run-sql` and `ci/circleci: validate-views` checks fail due to permissions error.

**Comment by kik-kik** [comment]
@jklukas @wlach could you please have a look if I missed anything? 🙇 

@wlach so looking at the query there are hard coded dates is this meant to be an incremental job or a full load?

File: sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/metadata.yaml
Comment (wlach) [inline sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/metadata.yaml:24]: I'm pretty sure this should not be `mozilla-confidential` since we want anyone to be able to access this particular table. @jklukas can give better guidance than me here though.

File: sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/query.sql
Comment (wlach) [inline sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/query.sql:16]: This should use the submission date parameter. See the example here: https://mozilla.github.io/bigquery-etl/cookbooks/creating_a_derived_dataset/#fill-out-the-query

**Review by wlach — CHANGES_REQUESTED** [review]
Looks pretty good to me overall, see comments inline.

This is a bit bikesheddy but we don't use the `Feat` terminology in our commit/PR messages:

I'd reword this:

`Feat(issues-2572): regrets_reporter_summary query (youtube)`

To something like:

`Add regrets reporter summary table (fixes #2572)`

The "fixes" part is useful, as it will autoresolve the issue when this merges.

**Comment by kik-kik** [comment]
Renamed the title as requested @wlach

**Review by wlach — APPROVED** [review]
r+wc. But wait for feedback from @jklukas (esp. re: permissions) before merging

**Comment by kik-kik** [comment]
Ok cool, all your suggestion @wlach should now be implemented. Waiting for input from @jklukas

File: sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/metadata.yaml
Comment (jklukas) [reply sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/metadata.yaml:27]: `workgroup:mozilla-confidential` corresponds to  "Mozilla Confidential - Staff and NDA'd Mozillians Only" per https://wiki.mozilla.org/Security/Data_Classification

It's the right thing here.

File: sql/moz-fx-data-shared-prod/regrets_reporter/regrets_reporter_summary/view.sql
Comment (jklukas) [inline sql/moz-fx-data-shared-prod/regrets_reporter/regrets_reporter_summary/view.sql:7]: We will also need a `metadata.yaml` in this view directory in order to add `mozilla-confidential` access to the view. Otherwise, it will be restricted at the dataset level.

File: sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/init.sql
Comment (jklukas) [inline sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/init.sql:102]:
```suggestion
    submission_timestamp >= "2021-12-02"
```

**Review by jklukas — APPROVED** [review]
r+wc (approval with comment)

I have not deeply reviewed the SQL, but the metadata here looks good aside from needing to add a `metadata.yaml` for the view.

**Review by jklukas — APPROVED** [review]
This also needs to have these queries added to the skip list in `bigquery_etl/dryrun.py` since the dry run service doesn't have privs to access the underlying tables.

**Comment by wlach** [comment]
> This also needs to have these queries added to the skip list in `bigquery_etl/dryrun.py` since the dry run service doesn't have privs to access the underlying tables.

Are there special deployment considerations in this case as well?

**Comment by jklukas** [comment]
> Are there special deployment considerations in this case as well?

We should ping @whd when this is completely ready and let him handle running the `init.sql`, applying privs, and merging the PR.

File: sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/init.sql
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/regrets_reporter_derived/regrets_reporter_summary_v1/init.sql:26]:
```suggestion
    submission_timestamp >= "2021-12-2"
```
```suggestion
    submission_timestamp >= "2021-12-02"
```

File: sql/moz-fx-data-shared-prod/regrets_reporter/regrets_reporter_summary/view.sql
Comment (kik-kik) [reply sql/moz-fx-data-shared-prod/regrets_reporter/regrets_reporter_summary/view.sql:7]: I added a `metadata.yaml` file in the same folder as the view definition. @jklukas could you please confirm this is what you meant and that the file is correct?

thanks!

**Comment by kik-kik** [comment]
@jklukas added regrets_reporter_summary to the exclusion list in `bigquery_etl/dryrun.py`. Let me know if you see any issues. 🙇

File: sql/moz-fx-data-shared-prod/regrets_reporter/regrets_reporter_summary/metadata.yaml
Comment (jklukas) [inline sql/moz-fx-data-shared-prod/regrets_reporter/regrets_reporter_summary/metadata.yaml:13]: This metadata looks good.

File: sql/moz-fx-data-shared-prod/regrets_reporter/regrets_reporter_summary/view.sql
Comment (jklukas) [reply sql/moz-fx-data-shared-prod/regrets_reporter/regrets_reporter_summary/view.sql:7]: Yes, looks good.

**Review by jklukas — APPROVED** [review]

**Comment by kik-kik** [comment]
> > Are there special deployment considerations in this case as well?
> 
> We should ping @whd when this is completely ready and let him handle running the `init.sql`, applying privs, and merging the PR.

@whd could you please run the `init.sql` and applying the correct settings on the table 🙏 , thanks! 🙇

**Comment by whd** [comment]
I ran `init.sql` and will apply permissions after this makes its way to the generated-sql branch.

**Comment by kik-kik** [comment]
Thank you @whd 🙇 

cc: @jklukas

**Pull request merged**
